psi4/driver/procrouting/empirical_dispersion.py

```diff
@@ -25,260 +25,243 @@
+                gcp_part = gcp.run_gcp(molecule, self.fctldash, verbose=False, dertype=0)
+                dashd_part += gcp_part
+
+            return dashd_part
         else:
-            return self.disp.compute_energy(molecule)
+            ene = self.disp.compute_energy(molecule)
+            core.set_variable('DISPERSION CORRECTION ENERGY', ene)
+            if self.fctldash:
+                core.set_variable('{} DISPERSION CORRECTION ENERGY'.format(self.fctldash), ene)
+            return ene
 
     def compute_gradient(self, molecule):
-        if self.disp_type == 'gr':
-            if self.alias in ['hf3c', 'pbeh3c']:
-                dashd_part = dftd3.run_dftd3(
-                    molecule,
-                    dashlvl=self.dtype.lower().replace('-', ''),
-                    dashparam=self.dash_params,
-                    verbose=False,
-                    dertype=1)
-                gcp_part = gcp.run_gcp(molecule, self.alias.lower(), verbose=False, dertype=1)
+        """Compute dispersion gradient based on engine, dispersion level, and parameters in `self`.
+
+        Parameters
+        ----------
+        molecule : psi4.core.Molecule
+            System for which to compute empirical dispersion correction.
+
+        Returns
+        -------
+        psi4.core.Matrix
+            (nat, 3) dispersion gradient [Eh/a0].
+
+        """
+        if self.engine == 'dftd3':
+            jobrec = intf_dftd3.run_dftd3_from_arrays(
+                molrec=molecule.to_dict(np_out=False),
+                name_hint=self.fctldash,
+                level_hint=self.dashlevel,
+                param_tweaks=self.dashparams,
+                dashcoeff_supplement=self.dashcoeff_supplement,
+                ptype='gradient',
+                verbose=1)
+
+            dashd_part = core.Matrix.from_array(jobrec['qcvars']['DISPERSION CORRECTION GRADIENT'].data)
+            for k, qca in jobrec['qcvars'].items():
+                if not isinstance(qca.data, np.ndarray):
+                    core.set_variable(k, qca.data)
+
+            if self.fctldash in ['hf3c', 'pbeh3c']:
+                gcp_part = gcp.run_gcp(molecule, self.fctldash, verbose=False, dertype=1)
                 dashd_part.add(gcp_part)
-                return dashd_part
-            else:
-                return dftd3.run_dftd3(
-                    molecule,
-                    dashlvl=self.dtype.lower().replace('-', ''),
-                    dashparam=self.dash_params,
-                    verbose=False,
-                    dertype=1)
+
+            return dashd_part
         else:
             return self.disp.compute_gradient(molecule)
 
     def compute_hessian(self, molecule):
-        """
-        #magic (if magic was easy)
-        """
+        """Compute dispersion Hessian based on engine, dispersion level, and parameters in `self`.
+        Uses finite difference, as no dispersion engine has analytic second derivatives.
```

Comment:
This presumably means that we won't (ever) have analytic 2nd derivatives for `-D3` functionals, as dftd3 doesn't support it?

Reply:
I haven't heard that analytic Hessians are on anyone's agenda. My impression is that findif H by G is pretty stable. But if any engine wants to support them, glad to change the EmpiricalDispersion class to accommodate. Is that what you were asking?

Reply:
No, but it answers my question. :+1: